### ISSUE TITLE
Docker build rename.  EXDEV: cross-device link not permitted

### ISSUE THREAD

**Issue by SC7639** [pnpm/pnpm#3439]
### Code to reproduce the issue:
.bitmap
```
/* THIS IS A BIT-AUTO-GENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */

{
    "version": "1.0.70"
}
```
package.json
```json 
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@material-ui/core": "^4.11.4"
  }
}
```
workspace.jsonc
``` jsonc 
/**
 * this is the main configuration file of your bit workspace.
 * for full documentation, please see: https://harmony-docs.bit.dev/workspace/configurations
 **/{
  "$schema": "https://static.bit.dev/teambit/schemas/schema.json",
  /**
   * main configuration of the Bit workspace.
   **/
  "teambit.workspace/workspace": {
    /**
     * the name of the component workspace. used for development purposes.
     **/
    "name": "my-workspace-name",
    /**
     * set the icon to be shown on the Bit server.
     **/
    "icon": "https://static.bit.dev/bit-logo.svg",
    /**
     * default directory for all components in workspace, the following placeholders are available:
     * name - component name includes namespace, e.g. 'ui/button'.
     * scopeId - full scope-id includes the owner, e.g. 'teambit.compilation'.
     * scope - scope name only, e.g. 'compilation'.
     * owner - owner name in bit.dev, e.g. 'teambit'.
     **/
    "defaultDirectory": "{scope}/{name}",
    /**
     * default scope for all components in workspace.
     **/
    "defaultScope": "my-scope"
  },
  /**
   * main configuration for component dependency resolution.
   **/
  "teambit.dependencies/dependency-resolver": {
    /**
     * choose the package manager for Bit to use. you can choose between 'yarn', 'pnpm'
     */
    "packageManager": "teambit.dependencies/pnpm",
    "policy": {
      "dependencies": {
        "@material-ui/core": "4.11.4"
      },
      "peerDependencies": {}
    },
    "packageManagerArgs": [],
    "devFilePatterns": [
      "**/*.spec.ts"
    ],
    "strictPeerDependencies": true
  },
  /**
   * workspace variants allow to set different subsets of configuration for components in your workspace.
   * this is extremely useful for upgrading, aligning and building components with a
   * new set of dependencies.
   **/
  "teambit.workspace/variants": {
    /**
     * "*" is a special rule which applied on all components in the workspace.
     **/
    "*": {
      /**
       * uncomment to apply the react environment on all components.
       **/
      // "teambit.react/react": { }
    }
  }
}
```

Dockerfile
```Dockerfile 
FROM node:14.15 as build

ADD package.json package.json
ADD workspace.jsonc workspace.jsonc
ADD .bitmap .bitmap
ADD .bit .bit


# Install bit command line tool and configure for ci/cd
RUN npx @teambit/bvm install
ENV PATH="/root/bin:${PATH}"
RUN echo "export PATH=${HOME}/bin:${PATH}" >> /root/.bashrc
RUN bit config set analytics_reporting false
RUN bit config set error_reporting false

# Install dependencies
RUN npm i


RUN bit install

CMD [ "tail", "-f", "/dev/null" ]

```
`docker build .`

### Expected behavior:
Expected pnpm renaming of files/folders into the `.ignore` directory to work in docker build 

### Actual behavior:
The issue is that when running `bit install` (using pnpm under the hood) after `npm install`, inside a Dockerfile build I got the following error, when pnpm was trying to move npm installed dependencies to `.ignored` directory using the `rename` command. 

```Dockerfile
RUN npm install
RUN bit install
```

```shell
EXDEV: cross-device link not permitted, rename '/app/node_modules/@material-ui/core' -> '/app/node_modules/.ignored/@material-ui/core'
```

Fixes: https://github.com/npm/npm/issues/9863
Explanation: https://blog.cloud66.com/using-node-with-docker/ (edited) 

tldr; Docker uses a filesystem that layers for each command (AUFS) and after each command is done the layer is read-only so the rename command can't be done as `node_modules` is in a ready only layer after `npm install` command completes. Move (mv) can handle the `EXDEV` error and works.

The fix I came up with is to do both `npm install` and `bit install` on the same `RUN` command then the `node_modules` directory is still in the writeable layer (top layer)

```Dockerfile
RUN npm install && bit install
```

### Additional information:

 - Windows, macOS, or Linux?: Docker build from `node:14.15` image

**Comment by SC7639** [comment]
I'll give that a go and get back to you.

**Comment by fatso83** [comment]
@SC7639 You never got back :smile_cat: How did it work?

**Comment by bobbyphilip** [comment]
I have this issue now.
I have Dockerfile with something as follows.
```
FROM node:16-alpine
RUN npm install -g pnpm
WORKDIR /app
COPY pnpm-lock.yaml ./
RUN pnpm fetch
COPY . ./
RUN pnpm install --offline
```  
pnpm fetch works,  and at the end of pnpm install, I get an error 

EXDEV  EXDEV: cross-device link not permitted, rename '/appbuild/mymodulename/node_modules/typescript' -> '/appbuild/mymodulename/node_modules/.ignored_typescript'

Strangely this only happens on my CI server, locally it works alright

**Comment by bobbyphilip** [comment]
I tried a variant of one of the old npm workarounds
RUN npm install -g fs-extra \
    && cd $(npm root -g)/pnpm \
  && sed -i -e s/graceful-fs/fs-extra/ -e s/fs\\.rename/fs.move/ ./dist/pnpm.cjs
but that alone didnt work  :-(

**Comment by SC7639** [comment]
> @SC7639 You never got back 😸 How did it work? Seeing the same issue running [Teamcity within Docker](https://hub.docker.com/r/jetbrains/teamcity-minimal-agent/).

Yes I did find a workaround it was to add a script entry to `package.json` -> `"postinstall": "rm -rf ./node_modules/.ignored"`

**Comment by elevensky** [comment]
> I have this issue now. I have Dockerfile with something as follows.
> 
> ```
> FROM node:16-alpine
> RUN npm install -g pnpm
> WORKDIR /app
> COPY pnpm-lock.yaml ./
> RUN pnpm fetch
> COPY . ./
> RUN pnpm install --offline
> ```
> 
> pnpm fetch works, and at the end of pnpm install, I get an error
> 
> EXDEV  EXDEV: cross-device link not permitted, rename '/appbuild/mymodulename/node_modules/typescript' -> '/appbuild/mymodulename/node_modules/.ignored_typescript'
> 
> Strangely this only happens on my CI server, locally it works alright

same issues, when set gitlab ci runner cache volumes

**Comment by brunojppb** [comment]
I was having this issue issues with version `7.13.0` but today I tried with version `7.16.0` and it worked like a charm.
Thanks folks for the fix!

**Comment by zkochan** [comment]
I believe we have fixed this in [v7.14.2](https://github.com/pnpm/pnpm/releases/tag/v7.14.2)

**Comment by elevensky** [comment]
> 

<img width="964" alt="image" src="https://user-images.githubusercontent.com/6904834/215969066-5d807c70-7b8a-4973-bb3c-ff73db0320d0.png">
in gitlab ci/cd docker runner still same, pnpm 7.24.3
